src/module_4/notebook.md

```diff
@@ -0,0 +1,2102 @@
+    roc_auc = auc(fpr, tpr)
+    ax[0].plot(fpr, tpr, label=f'{model_name} (AUC = {roc_auc:.2f})')
+    ax[0].set_xlabel('False Positive Rate')
+    ax[0].set_ylabel('True Positive Rate')
+    ax[0].set_title('ROC Curve')
+    ax[0].legend()
+
+    # Precision-Recall Curve
+    precision, recall, pr_th = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall, precision)
+    ax[1].plot(recall, precision, label=f'{model_name} (AUC = {pr_auc:.2f})')
+    ax[1].set_xlabel('Recall')
+    ax[1].set_ylabel('Precision')
+    ax[1].set_title('Precision-Recall Curve')
+    ax[1].legend()
+
+    # Highlight precision target (if provided)
+    if threshold_target is not None:
+        idx_0 = np.argmin(np.abs(roc_th - threshold_target))
+        ax[0].scatter(fpr[idx_0], tpr[idx_0])
+        idx_1 = np.argmin(np.abs(pr_th - threshold_target))
+        ax[1].scatter(recall[idx_1], precision[idx_1]) 
+```
+
+
+```python
+def auc_score(y_test: pd.Series, y_pred: pd.Series) -> Tuple[float, float]:
+    """
+    Returns the ROC and PR curves AUC score for a given set of true labels and predicted probabilities.
+
+    Args:
+        y_test (array-like): True labels.
+        y_preds (array-like): Predicted probabilities or scores.
+
+    Returns:
+        Tuple (float, float): ROC AUC score, PR curve AUC score.
+    """
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
+    roc_auc = auc(fpr, tpr)
+    precision_, recall_, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall_, precision_)
+    return roc_auc, pr_auc
+```
+
+### Baseline
+
+We first try our baseline without any ML involved, just using a single column, `global_popularity` to predict the outcome.
+
+
+```python
+plot_roc_and_pr_curves("Baseline", y_pred=val_df["global_popularity"], y_test=val_df[label_col])
+baseline_auc = auc_score(val_df[label_col], val_df["global_popularity"])
+```
+
+
+    
+![png](notebook_files/notebook_18_0.png)
+    
+
+
+### Model training
+
+
+```python
+def feature_label_split(df: pd.DataFrame, label_col: str) -> Tuple[pd.DataFrame, pd.Series]:
+    X = df.drop(columns=[label_col], axis=1)
+    y = df[label_col]
+    return X, y
+
+X_train, y_train = feature_label_split(train_df, label_col)
+X_val, y_val = feature_label_split(val_df, label_col)
+X_test, y_test = feature_label_split(test_df, label_col)
+```
+
+
+```python
+train_cols = numerical_cols + binary_cols
+```
+
+### Linear Model
+
+From the task of the previous module, we plot the best models for reference. We need to make sure our non-linear models (more complex models) improve on performance in order to justify the increase in model complexity.
+
+
+```python
+pr_auc_score_val = []
+roc_auc_scores_val = []
+
+fig1, ax1 = plt.subplots(1, 2, figsize=(12, 5))
+fig1.suptitle("Train metrics")
+
+fig2, ax2 = plt.subplots(1, 2, figsize=(12, 5))
+fig2.suptitle("Validation metrics")
+
+lrs = [
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l2', C=1e-6)),
```

Comment:
how have you chosen these params? from the prev notbeooks? they seem extremely low